src/charm_state.py

```diff
@@ -165,6 +162,54 @@ def _valid_storage_size_str(size: str) -> bool:
     return size[-3:] in valid_suffixes and size[:-3].isdigit()
 
 
+OPENSTACK_CLOUDS_YAML_CONFIG_NAME = "experimental-openstack-clouds-yaml"
+CLOUDS_YAML_PATH = Path(Path.home() / ".config/openstack/clouds.yaml")
+
+
+def _validate_openstack_cloud_config(cloud_config: dict) -> None:
```

Comment:
These functions have been moved from `openstack_manager.py` to remove circular dependency (not just in typing)

Reply:
We should look at this in a bit more detail, it doesn't belong here. It could go into a third module around openstack initialisation or something like that

Reply:
This has now been refactored into a separate module. Thank you!